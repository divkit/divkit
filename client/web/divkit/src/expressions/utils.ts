import type { EvalContext, EvalValue, EvalValueBase } from './eval';
import type { Node, Variable } from './ast';
import type { EvalTypes } from './eval';
import type { VariablesMap } from './eval';
import { walk } from './walk';
import { parseColor, ParsedColor } from '../utils/correctColor';
import { padLeft } from '../utils/padLeft';
import { hasBigInt, MAX_INT, MIN_INT, toBigInt } from './bigint';

export function valToInternal(val: EvalValue): EvalValue {
    if (val.type === 'url' || val.type === 'color') {
        return {
            type: 'string',
            value: val.value
        };
    }

    return val;
}

export function dateToString(date: Date): string {
    return date
        .toISOString()
        .replace('T', ' ')
        .replace(/\.\d{3}Z$/, '');
}

export function valToString(val: EvalValue): string {
    if (val.type === 'string') {
        return val.value;
    } else if (val.type === 'integer') {
        return String(val.value);
    } else if (val.type === 'number') {
        let res = String(val.value);

        if (!res.includes('.')) {
            if (res.includes('e')) {
                res = res.replace('e', '.0e');
            } else {
                res += '.0';
            }
        }

        res = res.replace(/e\+?/i, 'E');

        return res;
    } else if (val.type === 'boolean') {
        return val.value ? 'true' : 'false';
    } else if (val.type === 'datetime') {
        return dateToString(val.value);
    } else if (val.type === 'color') {
        return stringifyColor(safeConvertColor(val.value));
    } else if (val.type === 'url') {
        return val.value;
    } else if (val.type === 'dict') {
        return '<dict>';
    }

    // For purpose when new eval value types will be added
    throw new Error(`Unexpected type ${(val as EvalValueBase).type}`);
}

export function valToPreview(val: EvalValue): string {
    let res = valToString(val);

    if (val.type === 'string') {
        res = "'" +
            res
                .replace(/\\/g, '\\\\')
                .replace(/'/g, '\\\'') +
            "'";
    }

    return res;
}

export function typeToString(type: EvalTypes): string {
    if (type === 'datetime') {
        return 'DateTime';
    }
    return type.charAt(0).toUpperCase() + type.substring(1);
}

export function roundInteger(ctx: EvalContext, val: number | bigint): number | bigint {
    if (hasBigInt) {
        return toBigInt(val);
    }

    if (val < 0) {
        const res = -Math.floor(-val as number);
        if (res === 0) {
            // -0 => 0
            return 0;
        }
        return res;
    }

    return Math.floor(val as number);
}

export function checkIntegerOverflow(ctx: EvalContext, val: number | bigint): void {
    if (val < MIN_INT || val > MAX_INT) {
        throw new Error('Integer overflow.');
    }

    if (!hasBigInt && (val > Number.MAX_SAFE_INTEGER || val < Number.MIN_SAFE_INTEGER)) {
        ctx.safeIntegerOverflow = true;
    }
}

export function gatherVarsFromAst(ast: Node): string[] {
    const res: string[] = [];

    walk(ast, {
        Variable(node: Variable): void {
            res.push(node.id.name);
        }
    });

    return res;
}

export function evalError(msg: string, details: string): never {
    throw new Error(`Failed to evaluate [${msg}]. ${details}`);
}

export function containsUnsetVariables(ast: Node, variables: VariablesMap): boolean {
    let result = false;

    walk(ast, {
        Variable(node: Variable): void {
            if (!variables.has(node.id.name)) {
                result = true;
            }
        }
    });

    return result;
}

export function safeConvertColor(color: string): ParsedColor {
    const res = parseColor(color);

    if (res) {
        return res;
    }

    throw new Error('Unable to convert value to Color, expected format #AARRGGBB.');
}

export function stringifyColor(color: ParsedColor): string {
    return `#${[color.a, color.r, color.g, color.b].map(it => {
        if (it < 0 || it > 255) {
            throw new Error('Value out of range 0..1.');
        }

        return padLeft(Math.round(it).toString(16), 2);
    }).join('').toUpperCase()}`;
}

export function transformColorValue(color: string): string {
    return stringifyColor(safeConvertColor(color));
}
