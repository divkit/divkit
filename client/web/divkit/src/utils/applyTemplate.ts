import { wrapError, type WrappedError } from './wrapError';
import type { DivBase, TemplateContext } from '../../typings/common';
import type { DivBaseData } from '../types/base';

export function applyTemplate<T extends DivBase>(
    json: T,
    templateContext: TemplateContext,
    templates: Record<string, unknown>,
    logError: (error: WrappedError) => void
): {
    json: T;
    templateContext: TemplateContext;
};
export function applyTemplate(
    json: DivBaseData,
    templateContext: TemplateContext,
    templates: Record<string, unknown>,
    logError: (error: WrappedError) => void
): {
    json: DivBaseData;
    templateContext: TemplateContext;
} {
    const template = templates[json.type];

    if (!template) {
        logError(wrapError(new Error('No such template'), {
            additional: {
                template: json.type
            }
        }));

        return {
            json,
            templateContext
        };
    }

    let i;
    const newContext: TemplateContext = {};

    for (i in templateContext) {
        if (templateContext.hasOwnProperty(i)) {
            newContext[i] = templateContext[i];
        }
    }

    for (i in json) {
        if (i === 'type' || i === '__proto__') {
            continue;
        }

        if (json.hasOwnProperty(i)) {
            newContext[i] = json[i as keyof typeof json];
        }
    }

    function copyTemplated(base: any, extender: any) {
        const keys = Object.keys(extender).filter(key => key !== '__proto__');
        const simpleKeys = keys.filter(key => key.charAt(0) !== '$');
        const templateKeys = keys.filter(key => key.charAt(0) === '$');

        simpleKeys.forEach(key => {
            const item = extender[key];

            if (typeof item === 'object' && item !== null) {
                base[key] = Array.isArray(item) ? [] : {};
                copyTemplated(base[key], item);
            } else {
                base[key] = item;
            }
        });

        templateKeys.forEach(key => {
            const item = extender[key];

            const val = newContext[item];

            if (val !== undefined) {
                const prop = key.substring(1);
                base[prop] = val;
            }
        });

        return base;
    }

    const newJson = copyTemplated({}, template);

    for (i in json) {
        if (i === 'type' || i === '__proto__') {
            continue;
        }

        if (json.hasOwnProperty(i)) {
            newJson[i] = json[i as keyof typeof json];
        }
    }

    return {
        json: newJson,
        templateContext: newContext
    };
}
