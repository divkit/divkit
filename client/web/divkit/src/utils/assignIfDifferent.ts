import { isDeepEqual } from './isDeepEqual';

export function assignIfDifferent<A, B>(newVal: A, defaultVal: B): A | B {
    if (isDeepEqual(newVal, defaultVal)) {
        return defaultVal;
    }
    return newVal;
}
