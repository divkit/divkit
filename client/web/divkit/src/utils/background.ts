import type { AlignmentHorizontal, AlignmentVertical } from '../types/alignment';
import type {
    Background,
    GradientBackground,
    ImageBackground,
    SolidBackground,
    RadialBackground,
    RadialGradientCenter,
    GradientColorPoint
} from '../types/background';
import type { ImageScale } from '../types/imageScale';
import type { MaybeMissing } from '../expressions/json';
import { correctColor } from './correctColor';
import { htmlFilter } from './htmlFilter';
import { Truthy } from './truthy';
import { pxToEmWithUnits } from './pxToEm';

export function getBackground(bgs: MaybeMissing<Background>[], direction: 'ltr' | 'rtl'): {
    color?: string;
    image?: string;
    size?: string;
    position?: string;
} {
    if (bgs.length === 1 && bgs[0].type === 'solid') {
        return solidUnique({
            bg: bgs[0]
        });
    }

    const res = bgs
        .map(bg => {
            if (bg.type === 'solid') {
                return solid({
                    bg
                });
            } else if (bg.type === 'gradient') {
                return gradient({
                    bg
                });
            } else if (bg.type === 'image') {
                return image({
                    bg,
                    direction
                });
            } else if (bg.type === 'radial_gradient') {
                return radial({
                    bg
                });
            }
        })
        .filter(Truthy)
        .reverse()
        .reduce(function(acc, item) {
            acc.image.push(item.image);
            acc.size.push(item.size || 'auto');
            acc.position.push(item.pos || '50% 50%');

            return acc;
        }, {
            image: [] as string[],
            size: [] as string[],
            position: [] as string[]
        });

    return {
        image: res.image.join(','),
        size: res.size.join(','),
        position: res.position.join(',')
    };
}

function solid(opts: {
    bg: MaybeMissing<SolidBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} {
    const color = correctColor(opts.bg.color || 'transparent');

    return {
        size: undefined,
        pos: undefined,
        image: `linear-gradient(to bottom,${color},${color})`
    };
}

function solidUnique(opts: {
    bg: MaybeMissing<SolidBackground>;
}): {
    color: string;
    size: string;
    position: string;
} {
    const color = correctColor(opts.bg.color || 'transparent');

    return {
        color,
        size: 'auto',
        position: '50% 50%'
    };
}

function colorMapToList(colorMap: MaybeMissing<GradientColorPoint[]>): string | undefined {
    if (!colorMap.every(it => it.color && typeof it.position === 'number' && it.position >= 0 && it.position <= 1)) {
        return;
    }

    const colors = colorMap as {
        color: string;
        position: number;
    }[];

    const sortedColors = colors.sort((a, b) => {
        if (Math.abs(a.position - b.position) < 1e-6) {
            return 0;
        }
        return a.position - b.position;
    });

    return sortedColors
        .map(color => `${correctColor(color.color)} ${(color.position * 100).toFixed(2)}%`)
        .join(',');
}

function gradient(opts: {
    bg: MaybeMissing<GradientBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} | undefined {
    if (!Array.isArray(opts.bg?.colors) && !Array.isArray(opts.bg?.color_map)) {
        return;
    }

    const colors = opts.bg.colors?.filter(Truthy);
    if (!colors?.length && !opts.bg?.color_map) {
        return;
    }

    let image: string;
    if (opts.bg.color_map) {
        const list = colorMapToList(opts.bg.color_map);
        if (!list) {
            return;
        }

        image = 'linear-gradient(' +
            (90 - Number(opts.bg.angle || 0) + 'deg') +
            ',' +
            list +
            ')';
    } else {
        if (!colors) {
            return;
        }

        image = 'linear-gradient(' +
            (90 - Number(opts.bg.angle || 0) + 'deg') +
            ',' +
            colors
                .map(color => correctColor(color))
                .join(',') +
            ')';
    }

    return {
        size: undefined,
        pos: undefined,
        image
    };
}

const RELATIVE_SIZE_MAP = {
    nearest_corner: 'closest-corner',
    farthest_corner: 'farthest-corner',
    nearest_side: 'closest-side',
    farthest_side: 'farthest-side'
};

function radialCenterToCss(center: MaybeMissing<RadialGradientCenter> | undefined): string {
    if (center && typeof center === 'object' && 'type' in center && center.value !== undefined) {
        if (center.type === 'fixed') {
            return pxToEmWithUnits(center.value);
        } else if (center.type === 'relative') {
            return `${Number(center.value) * 100}%`;
        }
    }

    return '50%';
}

function radial(opts: {
    bg: MaybeMissing<RadialBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} | undefined {
    if (!Array.isArray(opts.bg?.colors) && !Array.isArray(opts.bg?.color_map)) {
        return;
    }

    const colors = opts.bg.colors?.filter(Truthy);
    if (!colors?.length && !opts.bg?.color_map) {
        return;
    }

    let list;
    if (opts.bg.color_map) {
        list = colorMapToList(opts.bg.color_map);
    } else if (colors) {
        list = colors
            .map(color => correctColor(color))
            .join(',');
    }
    if (!list) {
        return;
    }

    const sizeVal = opts.bg.radius;
    let size;
    if (sizeVal && typeof sizeVal === 'object' && 'type' in sizeVal && sizeVal.value !== undefined) {
        if (sizeVal.type === 'fixed') {
            size = pxToEmWithUnits(sizeVal.value);
        } else if (sizeVal.type === 'relative') {
            size = RELATIVE_SIZE_MAP[sizeVal.value];
        }
    }

    const centerX = radialCenterToCss(opts.bg.center_x);
    const centerY = radialCenterToCss(opts.bg.center_y);

    return {
        size: undefined,
        pos: undefined,
        image:
            'radial-gradient(' +
            `circle ${size || 'farthest-corner'} at ${centerX} ${centerY}` +
            ',' +
            list +
            ')'
    };
}

function image(opts: {
    bg: MaybeMissing<ImageBackground>;
    direction: 'ltr' | 'rtl'
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} | undefined {
    const image = opts.bg?.image_url;

    if (!image) {
        return;
    }

    return {
        size: imageSize(opts.bg.scale),
        pos: imagePos(opts.bg, opts.direction),
        image: 'url("' + htmlFilter(image) + '")'
    };
}

export function imageSize(scale?: ImageScale): string {
    if (scale === 'fit') {
        return 'contain';
    } else if (scale === 'stretch') {
        return 'fill';
    } else if (scale === 'no_scale') {
        return 'none';
    }

    // 'fill' and default
    return 'cover';
}

export function imagePos(obj: {
    content_alignment_horizontal?: AlignmentHorizontal;
    content_alignment_vertical?: AlignmentVertical;
}, direction: 'ltr' | 'rtl'): string {
    let hpos: string;
    let vpos: string;

    if (
        obj.content_alignment_horizontal === 'left' ||
        direction === 'ltr' && obj.content_alignment_horizontal === 'start' ||
        direction === 'rtl' && obj.content_alignment_horizontal === 'end'
    ) {
        hpos = '0%';
    } else if (
        obj.content_alignment_horizontal === 'right' ||
        direction === 'ltr' && obj.content_alignment_horizontal === 'end' ||
        direction === 'rtl' && obj.content_alignment_horizontal === 'start'
    ) {
        hpos = '100%';
    } else {
        hpos = '50%';
    }

    if (obj.content_alignment_vertical === 'top') {
        vpos = '0%';
    } else if (obj.content_alignment_vertical === 'bottom') {
        vpos = '100%';
    } else {
        vpos = '50%';
    }

    return hpos + ' ' + vpos;
}
