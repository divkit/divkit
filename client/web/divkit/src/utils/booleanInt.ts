import type { BooleanInt } from '../../typings/common';

export function booleanInt(val: BooleanInt): boolean {
    return val === true || val === 1;
}
