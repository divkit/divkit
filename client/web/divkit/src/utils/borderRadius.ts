import type { CornersRadius } from '../types/border';
import { pxToEm } from './pxToEm';

export function borderRadius(cornersRadius: CornersRadius, defaultRadius = 0, fontSize = 10): string {
    return [
        cornersRadius['top-left'],
        cornersRadius['top-right'],
        cornersRadius['bottom-right'],
        cornersRadius['bottom-left']
    ]
        .map(it => pxToEm((it || defaultRadius) / fontSize * 10))
        .join(' ');
}
