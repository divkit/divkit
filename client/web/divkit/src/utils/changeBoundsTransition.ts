import { cubicInOut } from 'svelte/easing';
import type { ChangeBoundsTransition } from '../types/base';
import { lerp } from './lerp';
import { isPrefersReducedMotion } from './isPrefersReducedMotion';
import { EASING } from './easing';

const DEFAULT_DURATION = 200;
const DEFAULT_DELAY = 0;

export interface TransitionProps {
    rootBbox: DOMRect;
    beforeBbox: DOMRect;
    afterBbox: DOMRect;
    transition: ChangeBoundsTransition;
}

export function changeBoundsTransition(node: HTMLElement, {
    rootBbox,
    beforeBbox,
    afterBbox,
    transition
}: TransitionProps) {
    return {
        delay: transition.start_delay ?? DEFAULT_DELAY,
        duration: isPrefersReducedMotion() ? 0 : (transition.duration ?? DEFAULT_DURATION),
        easing: (transition.interpolator && transition.interpolator in EASING) ?
            EASING[transition.interpolator] :
            cubicInOut,
        css: (t: number) => {
            return [
                `top:${lerp(beforeBbox.top, afterBbox.top, t) - rootBbox.top}px`,
                `left:${lerp(beforeBbox.left, afterBbox.left, t) - rootBbox.left}px`,
                `width:${lerp(beforeBbox.width, afterBbox.width, t)}px`,
                `height:${lerp(beforeBbox.height, afterBbox.height, t)}px`
            ].join(';');
        }
    };
}
