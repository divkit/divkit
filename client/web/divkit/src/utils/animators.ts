import type { AnimatorDirection, TypedValue } from '../../typings/common';
import type { MaybeMissing } from '../expressions/json';
import type { Animator } from '../types/base';
import type { Variable, VariableType } from '../../typings/variables';
import type { ExecAnyActionsFunc } from '../context/root';
import { stringifyColor } from '../expressions/utils';
import { clamp } from './clamp';
import { parseColor } from './correctColor';
import { correctNonNegativeNumber } from './correctNonNegativeNumber';
import { correctPositiveNumber } from './correctPositiveNumber';
import { getEasing } from './easing';
import { lerp } from './lerp';

function correctDirection(direction: string | undefined): AnimatorDirection | undefined {
    if (direction === 'normal' || direction === 'reverse' || direction === 'alternate' || direction === 'alternate_reverse') {
        return direction;
    }
}

export type CalcedAnimator = MaybeMissing<Animator & {
    start_value_typed: TypedValue;
    end_value_typed: TypedValue;
}>;

export interface AnimatorInstance {
    stop(): void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createAnimator(
    animator: CalcedAnimator,
    variableInstance: Variable<any, VariableType>,
    onEnd: () => void,
    execAnyActions: ExecAnyActionsFunc
): AnimatorInstance | undefined {
    const duration = correctPositiveNumber(animator.duration, 0);

    if (!duration || animator.type !== 'color_animator' && animator.type !== 'number_animator') {
        return;
    }

    const startValue = (animator.start_value_typed ? animator.start_value_typed.value : animator.start_value) ??
        variableInstance.getValue();
    const endValue = animator.end_value_typed ? animator.end_value_typed.value : animator.end_value;

    if (startValue === undefined || endValue === undefined) {
        return;
    }

    if (
        animator.type === 'color_animator' && (typeof startValue !== 'string' && startValue !== undefined || typeof endValue !== 'string') ||
        animator.type === 'number_animator' && (typeof startValue !== 'number' && startValue !== undefined || typeof endValue !== 'number')
    ) {
        return;
    }

    const parsedStart = animator.type === 'color_animator' && parseColor(startValue as string);
    const parsedEnd = animator.type === 'color_animator' && parseColor(endValue as string);
    if (animator.type === 'color_animator' && (!parsedStart || !parsedEnd)) {
        return;
    }

    const delay = correctNonNegativeNumber(animator.start_delay, 0);
    const easing = getEasing(animator.interpolator || 'linear');
    const direction = correctDirection(animator.direction) || 'normal';
    // eslint-disable-next-line no-nested-ternary
    const repeatCount = animator.repeat_count?.type === 'infinity' ?
        Infinity :
        (animator.repeat_count?.type === 'fixed' ? correctNonNegativeNumber(animator.repeat_count?.value, 1) : 1);

    let animationTime = 0;
    let localTime = performance.now();
    const maxTime = repeatCount === Infinity ? Infinity : repeatCount * duration + delay;

    function mix(t: number): string | number {
        if (animator.type === 'color_animator') {
            if (!parsedStart || !parsedEnd) {
                throw new Error('Missing start/end value');
            }
            return stringifyColor({
                a: clamp(lerp(parsedStart.a, parsedEnd.a, t), 0, 255),
                r: clamp(lerp(parsedStart.r, parsedEnd.r, t), 0, 255),
                g: clamp(lerp(parsedStart.g, parsedEnd.g, t), 0, 255),
                b: clamp(lerp(parsedStart.b, parsedEnd.b, t), 0, 255)
            });
        }
        return lerp(startValue as number, endValue as number, t);
    }

    function tick(now: number) {
        const diff = now - localTime;
        localTime = now;

        animationTime += diff;

        if (animationTime >= delay) {
            let iterationNumber = Math.floor((animationTime - delay) / duration);

            let t = (animationTime - delay - iterationNumber * duration) / duration;

            if (iterationNumber >= repeatCount) {
                iterationNumber = repeatCount - 1;
                t = 1;
            }

            let dir: 'normal' | 'reverse';
            if (
                direction === 'normal' ||
                (direction === 'alternate' && iterationNumber % 2 === 0) ||
                direction === 'alternate_reverse' && iterationNumber % 2 === 1
            ) {
                dir = 'normal';
            } else {
                dir = 'reverse';
            }

            if (dir === 'reverse') {
                t = 1 - t;
            }

            const value = mix(easing(t));
            variableInstance.setValue(value);
        }

        if (animationTime < maxTime) {
            timer = requestAnimationFrame(tick);
        } else {
            onEnd();
            execAnyActions(animator.end_actions);
        }
    }

    let timer = requestAnimationFrame(tick);

    return {
        stop() {
            cancelAnimationFrame(timer);
            execAnyActions(animator.cancel_actions);
            execAnyActions(animator.end_actions);
        }
    };
}
