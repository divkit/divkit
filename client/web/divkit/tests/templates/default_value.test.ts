import {
    describe,
    expect,
    test,
    vi
} from 'vitest';

import { applyTemplate } from '../../src/utils/applyTemplate';

describe('default_value', () => {
    test('test_invalid_value_presented_directly', () => {
        const json = require('../../../../../test_data/template_test_data/default_value/test_invalid_value_presented_directly.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_invalid_value_referenced_by_link', () => {
        const json = require('../../../../../test_data/template_test_data/default_value/test_invalid_value_referenced_by_link.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });
});
