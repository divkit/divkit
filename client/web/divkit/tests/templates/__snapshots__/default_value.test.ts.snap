// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default_value > test_invalid_value_presented_directly 1`] = `
{
  "json": {
    "int": -1,
    "nested": {
      "int": -1,
      "non_optional": "Some text",
      "url": "http://notsecure.yandex.ru",
    },
    "type": "entity_with_property_with_default_value",
    "url": "http://notsecure.yandex.ru",
  },
  "templateContext": {},
}
`;

exports[`default_value > test_invalid_value_presented_directly 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`default_value > test_invalid_value_referenced_by_link 1`] = `
{
  "json": {
    "int": -1,
    "int_value": -1,
    "nested": {
      "int": -1,
      "non_optional": "Some text",
      "url": "http://notsecure.yandex.ru",
    },
    "type": "entity_with_property_with_default_value",
    "url": "http://notsecure.yandex.ru",
    "url_value": "http://notsecure.yandex.ru",
  },
  "templateContext": {
    "int_value": -1,
    "url_value": "http://notsecure.yandex.ru",
  },
}
`;

exports[`default_value test_invalid_value_presented_directly 1`] = `
{
  "json": {
    "int": -1,
    "nested": {
      "int": -1,
      "non_optional": "Some text",
      "url": "http://notsecure.yandex.ru",
    },
    "type": "entity_with_property_with_default_value",
    "url": "http://notsecure.yandex.ru",
  },
  "templateContext": {},
}
`;

exports[`default_value test_invalid_value_presented_directly 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`default_value test_invalid_value_referenced_by_link 1`] = `
{
  "json": {
    "int": -1,
    "int_value": -1,
    "nested": {
      "int": -1,
      "non_optional": "Some text",
      "url": "http://notsecure.yandex.ru",
    },
    "type": "entity_with_property_with_default_value",
    "url": "http://notsecure.yandex.ru",
    "url_value": "http://notsecure.yandex.ru",
  },
  "templateContext": {
    "int_value": -1,
    "url_value": "http://notsecure.yandex.ru",
  },
}
`;
